File: code/packages/ruby/sql_parser/test/test_sql_parser.rb
# frozen_string_literal: true

require_relative "test_helper"

# ================================================================
# Tests for the SQL Parser
# ================================================================
#
# These tests verify that the grammar-driven parser, when loaded
# with sql.grammar, correctly builds Abstract Syntax Trees from
# ANSI SQL text.
#
# The grammar-driven parser produces generic ASTNode objects:
#
#   ASTNode(rule_name: "program", children: [...])
#
# Each node records which grammar rule produced it and its matched
# children (which can be tokens or other ASTNodes). This is different
# from a hand-written parser's typed nodes (SelectStatement, etc.),
# but it captures the same structural information.
#
# SQL's grammar (sql.grammar) has the following key rules:
#
#   program           = statement { ";" statement } [ ";" ]
#   statement         = select_stmt | insert_stmt | update_stmt
#                     | delete_stmt | create_table_stmt | drop_table_stmt
#   select_stmt       = "SELECT" [...] select_list "FROM" table_ref [...]
#   insert_stmt       = "INSERT" "INTO" NAME [...] "VALUES" row_value {...}
#   update_stmt       = "UPDATE" NAME "SET" assignment [...] [where_clause]
#   delete_stmt       = "DELETE" "FROM" NAME [where_clause]
#   create_table_stmt = "CREATE" "TABLE" [...] NAME "(" col_def {...} ")"
#   drop_table_stmt   = "DROP" "TABLE" [...] NAME
#
# Case insensitivity:
#   Because sql.tokens uses @case_insensitive true, keywords are
#   normalized to uppercase by the sql_lexer. So "select", "SELECT",
#   and "Select" all parse identically.
# ================================================================

class TestSqlParser < Minitest::Test
  ASTNode = CodingAdventures::Parser::ASTNode
  TT = CodingAdventures::Lexer::TokenType

  # SQL-specific token type strings
  NOT_EQUALS_TYPE    = "NOT_EQUALS"
  LESS_EQUALS_TYPE   = "LESS_EQUALS"
  GREATER_EQUALS_TYPE = "GREATER_EQUALS"

  # ------------------------------------------------------------------
  # Helpers
  # ------------------------------------------------------------------

  def parse(source)
    CodingAdventures::SqlParser.parse_sql(source)
  end

  # Recursively collect all Token objects from an AST.
  def collect_tokens(node)
    tokens = []
    return tokens unless node.is_a?(ASTNode)

    node.children.each do |child|
      if child.is_a?(CodingAdventures::Lexer::Token)
        tokens << child
      elsif child.is_a?(ASTNode)
        tokens.concat(collect_tokens(child))
      end
    end
    tokens
  end

  # Find all ASTNode descendants with a given rule_name.
  def find_nodes_by_rule(node, rule_name)
    results = []
    return results unless node.is_a?(ASTNode)

    results << node if node.rule_name == rule_name
    node.children.each do |child|
      results.concat(find_nodes_by_rule(child, rule_name)) if child.is_a?(ASTNode)
    end
    results
  end

  # Collect all keyword token values from an AST.
  def keyword_values(node)
    collect_tokens(node).select { |t| t.type == TT::KEYWORD }.map(&:value)
  end

  # ------------------------------------------------------------------
  # create_sql_parser
  # ------------------------------------------------------------------

  def test_create_sql_parser_returns_non_nil
    parser = CodingAdventures::SqlParser.create_sql_parser("SELECT 1 FROM t")
    refute_nil parser
  end

  def test_create_sql_parser_returns_grammar_driven_parser
    parser = CodingAdventures::SqlParser.create_sql_parser("SELECT 1 FROM t")
    assert_instance_of CodingAdventures::Parser::GrammarDrivenParser, parser
  end

  # ------------------------------------------------------------------
  # Grammar path
  # ------------------------------------------------------------------

  def test_grammar_path_exists
    assert File.exist?(CodingAdventures::SqlParser::SQL_GRAMMAR_PATH),
      "sql.grammar file should exist at #{CodingAdventures::SqlParser::SQL_GRAMMAR_PATH}"
  end

  # ------------------------------------------------------------------
  # Root node is always 'program'
  # ------------------------------------------------------------------
  # Every SQL parse tree starts with the 'program' rule (see sql.grammar
  # first line: program = statement { ";" statement } [ ";" ])

  def test_root_is_program
    ast = parse("SELECT 1 FROM t")
    assert_equal "program", ast.rule_name
  end

  # ------------------------------------------------------------------
  # SELECT statements
  # ------------------------------------------------------------------

  def test_simple_select
    # SELECT * FROM users
    ast = parse("SELECT * FROM users")
    assert_equal "program", ast.rule_name

    # Should contain a select_stmt node
    select_nodes = find_nodes_by_rule(ast, "select_stmt")
    assert select_nodes.length >= 1, "Expected a select_stmt node"

    # Should have SELECT and FROM keywords
    kw = keyword_values(ast)
    assert_includes kw, "SELECT"
    assert_includes kw, "FROM"
  end

  def test_select_with_column_list
    # SELECT id, name FROM users
    ast = parse("SELECT id, name FROM users")
    kw = keyword_values(ast)
    assert_includes kw, "SELECT"
    assert_includes kw, "FROM"

    all_tokens = collect_tokens(ast)
    name_tokens = all_tokens.select { |t| t.type == TT::NAME }
    name_values = name_tokens.map(&:value)
    assert_includes name_values, "id"
    assert_includes name_values, "name"
    assert_includes name_values, "users"
  end

  def test_select_with_where_clause
    # SELECT id FROM users WHERE id = 1
    ast = parse("SELECT id FROM users WHERE id = 1")
    kw = keyword_values(ast)
    assert_includes kw, "WHERE"

    where_nodes = find_nodes_by_rule(ast, "where_clause")
    assert where_nodes.length >= 1, "Expected a where_clause node"
  end

  def test_select_with_string_literal
    # SELECT * FROM users WHERE name = 'Alice'
    ast = parse("SELECT * FROM users WHERE name = 'Alice'")

    all_tokens = collect_tokens(ast)
    string_tokens = all_tokens.select { |t| t.type == TT::STRING }
    assert string_tokens.any? { |t| t.value == "Alice" }, "Expected STRING 'Alice'"
  end

  def test_select_with_number_literal
    ast = parse("SELECT * FROM orders WHERE id = 42")

    all_tokens = collect_tokens(ast)
    number_tokens = all_tokens.select { |t| t.type == TT::NUMBER }
    assert number_tokens.any? { |t| t.value == "42" }, "Expected NUMBER 42"
  end

  def test_select_star
    ast = parse("SELECT * FROM orders")
    select_list_nodes = find_nodes_by_rule(ast, "select_list")
    assert select_list_nodes.length >= 1, "Expected select_list node"

    all_tokens = collect_tokens(ast)
    star_tokens = all_tokens.select { |t| t.type == TT::STAR }
    assert star_tokens.any?, "Expected STAR token for SELECT *"
  end

  def test_select_with_limit
    ast = parse("SELECT id FROM users LIMIT 10")
    kw = keyword_values(ast)
    assert_includes kw, "LIMIT"

    limit_nodes = find_nodes_by_rule(ast, "limit_clause")
    assert limit_nodes.length >= 1, "Expected limit_clause node"
  end

  def test_select_with_order_by
    ast = parse("SELECT id FROM users ORDER BY id ASC")
    kw = keyword_values(ast)
    assert_includes kw, "ORDER"
    assert_includes kw, "BY"
    assert_includes kw, "ASC"
  end

  def test_select_with_group_by
    ast = parse("SELECT dept, COUNT(id) FROM employees GROUP BY dept")
    kw = keyword_values(ast)
    assert_includes kw, "GROUP"
    assert_includes kw, "BY"
  end

  def test_select_distinct
    ast = parse("SELECT DISTINCT name FROM users")
    kw = keyword_values(ast)
    assert_includes kw, "DISTINCT"
  end

  def test_select_with_as_alias
    ast = parse("SELECT id AS user_id FROM users")
    kw = keyword_values(ast)
    assert_includes kw, "AS"
  end

  def test_select_with_and_condition
    ast = parse("SELECT * FROM t WHERE a = 1 AND b = 2")
    kw = keyword_values(ast)
    assert_includes kw, "AND"
  end

  def test_select_with_or_condition
    ast = parse("SELECT * FROM t WHERE a = 1 OR b = 2")
    kw = keyword_values(ast)
    assert_includes kw, "OR"
  end

  def test_select_with_not_equals
    ast = parse("SELECT * FROM t WHERE id != 0")
    all_tokens = collect_tokens(ast)
    ne_tokens = all_tokens.select { |t| t.type == NOT_EQUALS_TYPE }
    assert ne_tokens.any?, "Expected NOT_EQUALS token"
  end

  def test_select_with_less_equals
    ast = parse("SELECT * FROM t WHERE price <= 100")
    all_tokens = collect_tokens(ast)
    le_tokens = all_tokens.select { |t| t.type == LESS_EQUALS_TYPE }
    assert le_tokens.any?, "Expected LESS_EQUALS token"
  end

  def test_select_with_greater_equals
    ast = parse("SELECT * FROM t WHERE score >= 90")
    all_tokens = collect_tokens(ast)
    ge_tokens = all_tokens.select { |t| t.type == GREATER_EQUALS_TYPE }
    assert ge_tokens.any?, "Expected GREATER_EQUALS token"
  end

  def test_select_with_is_null
    ast = parse("SELECT * FROM users WHERE email IS NULL")
    kw = keyword_values(ast)
    assert_includes kw, "IS"
    assert_includes kw, "NULL"
  end

  def test_select_with_is_not_null
    ast = parse("SELECT * FROM users WHERE email IS NOT NULL")
    kw = keyword_values(ast)
    assert_includes kw, "IS"
    assert_includes kw, "NOT"
    assert_includes kw, "NULL"
  end

  def test_select_with_in_clause
    ast = parse("SELECT * FROM t WHERE id IN (1, 2, 3)")
    kw = keyword_values(ast)
    assert_includes kw, "IN"
  end

  def test_select_with_between
    ast = parse("SELECT * FROM products WHERE price BETWEEN 10 AND 100")
    kw = keyword_values(ast)
    assert_includes kw, "BETWEEN"
    assert_includes kw, "AND"
  end

  def test_select_with_like
    ast = parse("SELECT * FROM users WHERE name LIKE 'A%'")
    kw = keyword_values(ast)
    assert_includes kw, "LIKE"
  end

  def test_select_with_qualified_column
    # SELECT t.id FROM t
    ast = parse("SELECT t.id FROM t")
    all_tokens = collect_tokens(ast)
    dot_tokens = all_tokens.select { |t| t.type == TT::DOT }
    assert dot_tokens.any?, "Expected DOT token for qualified column t.id"
  end

  def test_select_function_call
    # SELECT COUNT(*) FROM t
    ast = parse("SELECT COUNT(id) FROM t")

    function_nodes = find_nodes_by_rule(ast, "function_call")
    assert function_nodes.length >= 1, "Expected function_call node for COUNT(id)"
  end

  def test_select_with_join
    # The grammar requires join_type before JOIN: INNER JOIN, LEFT JOIN, etc.
    # Plain JOIN without a qualifier is not in the grammar.
    ast = parse("SELECT u.id, o.total FROM users AS u INNER JOIN orders AS o ON u.id = o.user_id")
    kw = keyword_values(ast)
    assert_includes kw, "INNER"
    assert_includes kw, "JOIN"
    assert_includes kw, "ON"
    assert_includes kw, "AS"
  end

  def test_select_with_left_join
    ast = parse("SELECT * FROM users LEFT JOIN orders ON users.id = orders.user_id")
    kw = keyword_values(ast)
    assert_includes kw, "LEFT"
    assert_includes kw, "JOIN"
    assert_includes kw, "ON"
  end

  def test_select_with_true_false
    ast = parse("SELECT * FROM users WHERE active = TRUE")
    kw = keyword_values(ast)
    assert_includes kw, "TRUE"
  end

  # ------------------------------------------------------------------
  # INSERT statements
  # ------------------------------------------------------------------

  def test_insert_statement
    # INSERT INTO users (name, email) VALUES ('Alice', 'alice@example.com')
    ast = parse("INSERT INTO users (name, email) VALUES ('Alice', 'alice@example.com')")
    assert_equal "program", ast.rule_name

    insert_nodes = find_nodes_by_rule(ast, "insert_stmt")
    assert insert_nodes.length >= 1, "Expected an insert_stmt node"

    kw = keyword_values(ast)
    assert_includes kw, "INSERT"
    assert_includes kw, "INTO"
    assert_includes kw, "VALUES"
  end

  def test_insert_values_contain_strings
    ast = parse("INSERT INTO users VALUES ('Bob')")
    all_tokens = collect_tokens(ast)
    string_tokens = all_tokens.select { |t| t.type == TT::STRING }
    assert string_tokens.any? { |t| t.value == "Bob" }
  end

  def test_insert_without_column_list
    # INSERT INTO ... VALUES (...) without specifying column names
    ast = parse("INSERT INTO orders VALUES (1, 99.99)")
    kw = keyword_values(ast)
    assert_includes kw, "INSERT"
    assert_includes kw, "VALUES"
  end

  # ------------------------------------------------------------------
  # UPDATE statements
  # ------------------------------------------------------------------

  def test_update_statement
    # UPDATE users SET name = 'Bob' WHERE id = 1
    ast = parse("UPDATE users SET name = 'Bob' WHERE id = 1")
    assert_equal "program", ast.rule_name

    update_nodes = find_nodes_by_rule(ast, "update_stmt")
    assert update_nodes.length >= 1, "Expected an update_stmt node"

    kw = keyword_values(ast)
    assert_includes kw, "UPDATE"
    assert_includes kw, "SET"
    assert_includes kw, "WHERE"
  end

  def test_update_multiple_assignments
    ast = parse("UPDATE users SET name = 'Bob', age = 30 WHERE id = 1")
    assignment_nodes = find_nodes_by_rule(ast, "assignment")
    assert_equal 2, assignment_nodes.length, "Expected 2 assignment nodes"
  end

  def test_update_without_where
    ast = parse("UPDATE counters SET value = 0")
    kw = keyword_values(ast)
    assert_includes kw, "UPDATE"
    assert_includes kw, "SET"
    refute_includes kw, "WHERE", "Should have no WHERE keyword"
  end

  # ------------------------------------------------------------------
  # DELETE statements
  # ------------------------------------------------------------------

  def test_delete_statement
    # DELETE FROM orders WHERE id = 42
    ast = parse("DELETE FROM orders WHERE id = 42")
    assert_equal "program", ast.rule_name

    delete_nodes = find_nodes_by_rule(ast, "delete_stmt")
    assert delete_nodes.length >= 1, "Expected a delete_stmt node"

    kw = keyword_values(ast)
    assert_includes kw, "DELETE"
    assert_includes kw, "FROM"
    assert_includes kw, "WHERE"
  end

  def test_delete_without_where
    ast = parse("DELETE FROM temp_data")
    kw = keyword_values(ast)
    assert_includes kw, "DELETE"
    assert_includes kw, "FROM"
  end

  # ------------------------------------------------------------------
  # CREATE TABLE statements
  # ------------------------------------------------------------------

  def test_create_table_statement
    sql = "CREATE TABLE users (id INTEGER, name VARCHAR)"
    ast = parse(sql)
    assert_equal "program", ast.rule_name

    create_nodes = find_nodes_by_rule(ast, "create_table_stmt")
    assert create_nodes.length >= 1, "Expected a create_table_stmt node"

    kw = keyword_values(ast)
    assert_includes kw, "CREATE"
    assert_includes kw, "TABLE"
  end

  def test_create_table_with_column_constraints
    sql = "CREATE TABLE products (id INTEGER NOT NULL, name TEXT NOT NULL, price DECIMAL DEFAULT 0)"
    ast = parse(sql)
    kw = keyword_values(ast)
    assert_includes kw, "CREATE"
    assert_includes kw, "TABLE"
    assert_includes kw, "NOT"
    assert_includes kw, "NULL"
    assert_includes kw, "DEFAULT"
  end

  def test_create_table_if_not_exists
    sql = "CREATE TABLE IF NOT EXISTS sessions (id INTEGER)"
    ast = parse(sql)
    kw = keyword_values(ast)
    assert_includes kw, "IF"
    assert_includes kw, "NOT"
    assert_includes kw, "EXISTS"
  end

  def test_create_table_with_primary_key
    sql = "CREATE TABLE users (id INTEGER PRIMARY KEY)"
    ast = parse(sql)
    kw = keyword_values(ast)
    assert_includes kw, "PRIMARY"
    assert_includes kw, "KEY"
  end

  # ------------------------------------------------------------------
  # DROP TABLE statements
  # ------------------------------------------------------------------

  def test_drop_table_statement
    ast = parse("DROP TABLE orders")
    assert_equal "program", ast.rule_name

    drop_nodes = find_nodes_by_rule(ast, "drop_table_stmt")
    assert drop_nodes.length >= 1, "Expected a drop_table_stmt node"

    kw = keyword_values(ast)
    assert_includes kw, "DROP"
    assert_includes kw, "TABLE"
  end

  def test_drop_table_if_exists
    ast = parse("DROP TABLE IF EXISTS temp_table")
    kw = keyword_values(ast)
    assert_includes kw, "IF"
    assert_includes kw, "EXISTS"
  end

  # ------------------------------------------------------------------
  # Multiple statements (separated by semicolons)
  # ------------------------------------------------------------------

  def test_multiple_statements
    ast = parse("SELECT 1 FROM a; SELECT 2 FROM b")
    assert_equal "program", ast.rule_name

    select_nodes = find_nodes_by_rule(ast, "select_stmt")
    assert_equal 2, select_nodes.length, "Expected 2 select_stmt nodes"
  end

  def test_multiple_statements_with_trailing_semicolon
    ast = parse("SELECT 1 FROM a;")
    assert_equal "program", ast.rule_name
    # The trailing semicolon should parse without error
  end

  # ------------------------------------------------------------------
  # Case-insensitive keywords
  # ------------------------------------------------------------------
  # Because sql.tokens has @case_insensitive true, the sql_lexer
  # normalizes keyword values to uppercase. The grammar matches
  # uppercase keyword literals, so lowercase input parses identically.

  def test_lowercase_select_parses
    ast = parse("select id from users")
    assert_equal "program", ast.rule_name
    kw = keyword_values(ast)
    assert_includes kw, "SELECT", "lowercase 'select' should be normalized to 'SELECT'"
    assert_includes kw, "FROM"
  end

  def test_mixed_case_keywords_parse
    ast = parse("Select Id From Users Where Id = 1")
    assert_equal "program", ast.rule_name
    kw = keyword_values(ast)
    assert_includes kw, "SELECT"
    assert_includes kw, "FROM"
    assert_includes kw, "WHERE"
  end

  def test_all_lowercase_insert_parses
    ast = parse("insert into users values ('Alice')")
    assert_equal "program", ast.rule_name
    kw = keyword_values(ast)
    assert_includes kw, "INSERT"
    assert_includes kw, "INTO"
    assert_includes kw, "VALUES"
  end

  # ------------------------------------------------------------------
  # Expressions
  # ------------------------------------------------------------------

  def test_arithmetic_expression_in_select
    ast = parse("SELECT price * 1.1 FROM products")
    all_tokens = collect_tokens(ast)
    star_tokens = all_tokens.select { |t| t.type == TT::STAR }
    assert star_tokens.any?, "Expected STAR token for multiplication"
  end

  def test_nested_expression_with_parens
    ast = parse("SELECT * FROM t WHERE (a = 1 OR b = 2) AND c = 3")
    kw = keyword_values(ast)
    assert_includes kw, "OR"
    assert_includes kw, "AND"
  end

  def test_not_expression
    ast = parse("SELECT * FROM t WHERE NOT active = TRUE")
    kw = keyword_values(ast)
    assert_includes kw, "NOT"
  end

  # ------------------------------------------------------------------
  # Error cases
  # ------------------------------------------------------------------

  def test_error_on_empty_input
    assert_raises(StandardError) { parse("") }
  end

  def test_error_on_select_without_from
    # "SELECT *" alone without FROM is incomplete per the grammar
    assert_raises(StandardError) { parse("SELECT *") }
  end

  def test_error_on_unknown_token
    assert_raises(StandardError) { parse("@@invalid") }
  end
end

File: code/packages/lua/sql_parser/tests/test_sql_parser.lua
-- Tests for sql_parser
-- =====================
--
-- Comprehensive busted test suite for the SQL parser package.
--
-- Test coverage:
--   - Module loads and exposes the public API
--   - SELECT * FROM users → program → statement → select_stmt
--   - SELECT with column list and WHERE clause
--   - SELECT with DISTINCT
--   - INSERT INTO ... VALUES (...)
--   - UPDATE ... SET ... WHERE ...
--   - DELETE FROM ... WHERE ...
--   - Expressions: comparisons, AND/OR, arithmetic
--   - Grammar first rule is "program"
--   - create_parser returns a GrammarParser
--   - get_grammar returns a grammar with rules
--   - Invalid SQL raises an error

-- Resolve sibling packages from the monorepo so busted can find them
-- without requiring a global luarocks install.
package.path = (
    "../src/?.lua;"                                           ..
    "../src/?/init.lua;"                                      ..
    "../../grammar_tools/src/?.lua;"                          ..
    "../../grammar_tools/src/?/init.lua;"                     ..
    "../../lexer/src/?.lua;"                                  ..
    "../../lexer/src/?/init.lua;"                             ..
    "../../state_machine/src/?.lua;"                          ..
    "../../state_machine/src/?/init.lua;"                     ..
    "../../directed_graph/src/?.lua;"                         ..
    "../../directed_graph/src/?/init.lua;"                    ..
    "../../sql_lexer/src/?.lua;"                              ..
    "../../sql_lexer/src/?/init.lua;"                         ..
    "../../parser/src/?.lua;"                                 ..
    "../../parser/src/?/init.lua;"                            ..
    package.path
)

local sql_parser = require("coding_adventures.sql_parser")

-- =========================================================================
-- Helper utilities
-- =========================================================================

--- Recursively collect all rule_names from an AST, in pre-order.
-- @param node  ASTNode  The root node to walk.
-- @return table         Ordered list of rule_name strings encountered.
local function collect_rule_names(node)
    local out = {}
    local function walk(n)
        if type(n) ~= "table" then return end
        if n.rule_name then
            out[#out + 1] = n.rule_name
            if n.children then
                for _, child in ipairs(n.children) do
                    walk(child)
                end
            end
        end
    end
    walk(node)
    return out
end

--- Find the first ASTNode with the given rule_name (depth-first).
-- @param node      ASTNode  Root to search.
-- @param rule_name string   Rule name to find.
-- @return ASTNode|nil
local function find_node(node, rule_name)
    if type(node) ~= "table" then return nil end
    if node.rule_name == rule_name then return node end
    if node.children then
        for _, child in ipairs(node.children) do
            local found = find_node(child, rule_name)
            if found then return found end
        end
    end
    return nil
end

--- Count all nodes with the given rule_name (full traversal).
-- @param node      ASTNode
-- @param rule_name string
-- @return number
local function count_nodes(node, rule_name)
    if type(node) ~= "table" then return 0 end
    local n = (node.rule_name == rule_name) and 1 or 0
    if node.children then
        for _, child in ipairs(node.children) do
            n = n + count_nodes(child, rule_name)
        end
    end
    return n
end

-- =========================================================================
-- Module surface
-- =========================================================================

describe("sql_parser module", function()
    it("loads successfully", function()
        assert.is_not_nil(sql_parser)
    end)

    it("exposes VERSION as a string", function()
        assert.is_string(sql_parser.VERSION)
        assert.matches("^%d+%.%d+%.%d+$", sql_parser.VERSION)
    end)

    it("exposes parse as a function", function()
        assert.is_function(sql_parser.parse)
    end)

    it("exposes create_parser as a function", function()
        assert.is_function(sql_parser.create_parser)
    end)

    it("exposes get_grammar as a function", function()
        assert.is_function(sql_parser.get_grammar)
    end)

    it("get_grammar returns an object with rules", function()
        local g = sql_parser.get_grammar()
        assert.is_not_nil(g)
        assert.is_table(g.rules)
        -- sql.grammar has many rules (program, statement, select_stmt, etc.)
        assert.is_true(#g.rules >= 5)
    end)

    it("grammar first rule is program", function()
        local g = sql_parser.get_grammar()
        assert.are.equal("program", g.rules[1].name)
    end)
end)

-- =========================================================================
-- Root node
-- =========================================================================

describe("root AST node", function()
    it("parse returns a non-nil value", function()
        local ast = sql_parser.parse("SELECT * FROM users")
        assert.is_not_nil(ast)
    end)

    it("root node has rule_name == 'program'", function()
        local ast = sql_parser.parse("SELECT * FROM users")
        assert.are.equal("program", ast.rule_name)
    end)

    it("root node has children", function()
        local ast = sql_parser.parse("SELECT * FROM users")
        assert.is_table(ast.children)
        assert.is_true(#ast.children >= 1)
    end)

    it("program contains a statement node", function()
        local ast = sql_parser.parse("SELECT * FROM users")
        local stmt = find_node(ast, "statement")
        assert.is_not_nil(stmt, "expected 'statement' node")
    end)
end)

-- =========================================================================
-- SELECT statements
-- =========================================================================

describe("SELECT statements", function()
    it("parses SELECT * FROM users", function()
        local ast = sql_parser.parse("SELECT * FROM users")
        assert.are.equal("program", ast.rule_name)
        local sel = find_node(ast, "select_stmt")
        assert.is_not_nil(sel, "expected 'select_stmt' node")
    end)

    it("select_list node present for SELECT *", function()
        local ast = sql_parser.parse("SELECT * FROM users")
        local sl = find_node(ast, "select_list")
        assert.is_not_nil(sl, "expected 'select_list' node")
    end)

    it("table_ref node present", function()
        local ast = sql_parser.parse("SELECT * FROM users")
        local tr = find_node(ast, "table_ref")
        assert.is_not_nil(tr, "expected 'table_ref' node")
    end)

    it("parses SELECT with column list", function()
        local ast = sql_parser.parse("SELECT name, age FROM users")
        assert.are.equal("program", ast.rule_name)
        local sel = find_node(ast, "select_stmt")
        assert.is_not_nil(sel)
        -- Two select_item nodes expected
        local count = count_nodes(ast, "select_item")
        assert.is_true(count >= 2, "expected at least 2 select_item nodes")
    end)

    it("parses SELECT name, age FROM users WHERE age > 18", function()
        local ast = sql_parser.parse("SELECT name, age FROM users WHERE age > 18")
        assert.are.equal("program", ast.rule_name)
        local wc = find_node(ast, "where_clause")
        assert.is_not_nil(wc, "expected 'where_clause' node")
    end)

    it("parses SELECT DISTINCT", function()
        local ast = sql_parser.parse("SELECT DISTINCT name FROM users")
        assert.are.equal("program", ast.rule_name)
        local sel = find_node(ast, "select_stmt")
        assert.is_not_nil(sel)
    end)

    it("parses SELECT with AND condition", function()
        local ast = sql_parser.parse(
            "SELECT * FROM employees WHERE salary > 50000 AND active = TRUE"
        )
        assert.are.equal("program", ast.rule_name)
        local wc = find_node(ast, "where_clause")
        assert.is_not_nil(wc)
    end)

    it("parses SELECT with ORDER BY", function()
        local ast = sql_parser.parse(
            "SELECT * FROM users ORDER BY name"
        )
        assert.are.equal("program", ast.rule_name)
        local oc = find_node(ast, "order_clause")
        assert.is_not_nil(oc, "expected 'order_clause' node")
    end)

    it("parses SELECT with LIMIT", function()
        local ast = sql_parser.parse(
            "SELECT * FROM users LIMIT 10"
        )
        assert.are.equal("program", ast.rule_name)
        local lc = find_node(ast, "limit_clause")
        assert.is_not_nil(lc, "expected 'limit_clause' node")
    end)

    it("parses SELECT with JOIN", function()
        local ast = sql_parser.parse(
            "SELECT * FROM users INNER JOIN orders ON users.id = orders.user_id"
        )
        assert.are.equal("program", ast.rule_name)
        local jc = find_node(ast, "join_clause")
        assert.is_not_nil(jc, "expected 'join_clause' node")
    end)

    it("parses SELECT with GROUP BY", function()
        local ast = sql_parser.parse(
            "SELECT department FROM employees GROUP BY department"
        )
        assert.are.equal("program", ast.rule_name)
        local gc = find_node(ast, "group_clause")
        assert.is_not_nil(gc, "expected 'group_clause' node")
    end)
end)

-- =========================================================================
-- INSERT statements
-- =========================================================================

describe("INSERT statements", function()
    it("parses INSERT INTO orders VALUES (1, 'item', 9.99)", function()
        local ast = sql_parser.parse("INSERT INTO orders VALUES (1, 'item', 9.99)")
        assert.are.equal("program", ast.rule_name)
        local ins = find_node(ast, "insert_stmt")
        assert.is_not_nil(ins, "expected 'insert_stmt' node")
    end)

    it("insert_stmt has row_value node", function()
        local ast = sql_parser.parse("INSERT INTO t VALUES (42, 'hello')")
        local rv = find_node(ast, "row_value")
        assert.is_not_nil(rv, "expected 'row_value' node")
    end)

    it("parses INSERT with column list", function()
        local ast = sql_parser.parse(
            "INSERT INTO users (id, name) VALUES (1, 'Alice')"
        )
        assert.are.equal("program", ast.rule_name)
        local ins = find_node(ast, "insert_stmt")
        assert.is_not_nil(ins)
    end)

    it("parses INSERT with multiple values", function()
        local ast = sql_parser.parse(
            "INSERT INTO t VALUES (1, 2, 3, 4)"
        )
        assert.are.equal("program", ast.rule_name)
        local rv = find_node(ast, "row_value")
        assert.is_not_nil(rv)
    end)
end)

-- =========================================================================
-- UPDATE statements
-- =========================================================================

describe("UPDATE statements", function()
    it("parses UPDATE users SET name = 'Bob' WHERE id = 1", function()
        local ast = sql_parser.parse("UPDATE users SET name = 'Bob' WHERE id = 1")
        assert.are.equal("program", ast.rule_name)
        local upd = find_node(ast, "update_stmt")
        assert.is_not_nil(upd, "expected 'update_stmt' node")
    end)

    it("update_stmt has assignment node", function()
        local ast = sql_parser.parse("UPDATE t SET col = 42")
        local assign = find_node(ast, "assignment")
        assert.is_not_nil(assign, "expected 'assignment' node")
    end)

    it("parses UPDATE with WHERE clause", function()
        local ast = sql_parser.parse("UPDATE t SET x = 1 WHERE y = 2")
        local wc = find_node(ast, "where_clause")
        assert.is_not_nil(wc, "expected 'where_clause' node")
    end)

    it("parses UPDATE with multiple assignments", function()
        local ast = sql_parser.parse(
            "UPDATE users SET name = 'Alice', age = 30 WHERE id = 1"
        )
        assert.are.equal("program", ast.rule_name)
        local assign_count = count_nodes(ast, "assignment")
        assert.is_true(assign_count >= 2, "expected at least 2 assignment nodes")
    end)
end)

-- =========================================================================
-- DELETE statements
-- =========================================================================

describe("DELETE statements", function()
    it("parses DELETE FROM temp WHERE expired = TRUE", function()
        local ast = sql_parser.parse("DELETE FROM temp WHERE expired = TRUE")
        assert.are.equal("program", ast.rule_name)
        local del = find_node(ast, "delete_stmt")
        assert.is_not_nil(del, "expected 'delete_stmt' node")
    end)

    it("parses DELETE without WHERE clause", function()
        local ast = sql_parser.parse("DELETE FROM temp")
        assert.are.equal("program", ast.rule_name)
        local del = find_node(ast, "delete_stmt")
        assert.is_not_nil(del)
    end)

    it("delete_stmt has where_clause when present", function()
        local ast = sql_parser.parse("DELETE FROM t WHERE id = 5")
        local wc = find_node(ast, "where_clause")
        assert.is_not_nil(wc)
    end)
end)

-- =========================================================================
-- Expression nodes
-- =========================================================================

describe("expression nodes", function()
    it("comparison produces comparison node", function()
        local ast = sql_parser.parse("SELECT * FROM t WHERE a > 1")
        local cmp = find_node(ast, "comparison")
        assert.is_not_nil(cmp, "expected 'comparison' node")
    end)

    it("column_ref node present in column list", function()
        local ast = sql_parser.parse("SELECT id FROM users")
        local cr = find_node(ast, "column_ref")
        assert.is_not_nil(cr, "expected 'column_ref' node")
    end)

    it("or_expr present in nested WHERE condition", function()
        local ast = sql_parser.parse(
            "SELECT * FROM t WHERE a > 1 OR b < 2"
        )
        local oe = find_node(ast, "or_expr")
        assert.is_not_nil(oe, "expected 'or_expr' node")
    end)

    it("and_expr present in AND WHERE condition", function()
        local ast = sql_parser.parse(
            "SELECT * FROM t WHERE a > 1 AND b < 2"
        )
        local ae = find_node(ast, "and_expr")
        assert.is_not_nil(ae, "expected 'and_expr' node")
    end)

    it("additive node present in arithmetic expression", function()
        local ast = sql_parser.parse("SELECT a + b FROM t")
        local add = find_node(ast, "additive")
        assert.is_not_nil(add, "expected 'additive' node")
    end)

    it("function_call node for COUNT(*)", function()
        local ast = sql_parser.parse(
            "SELECT COUNT(*) FROM employees"
        )
        local fc = find_node(ast, "function_call")
        assert.is_not_nil(fc, "expected 'function_call' node")
    end)
end)

-- =========================================================================
-- Semicolon-separated multiple statements
-- =========================================================================

describe("multiple statements", function()
    it("parses two statements separated by semicolon", function()
        local ast = sql_parser.parse(
            "SELECT * FROM a; SELECT * FROM b"
        )
        assert.are.equal("program", ast.rule_name)
        local count = count_nodes(ast, "select_stmt")
        assert.are.equal(2, count)
    end)

    it("parses statement with trailing semicolon", function()
        local ast = sql_parser.parse("SELECT 1 FROM t;")
        assert.are.equal("program", ast.rule_name)
        local sel = find_node(ast, "select_stmt")
        assert.is_not_nil(sel)
    end)
end)

-- =========================================================================
-- create_parser
-- =========================================================================

describe("create_parser", function()
    it("returns a non-nil parser object", function()
        local p = sql_parser.create_parser("SELECT * FROM t")
        assert.is_not_nil(p)
    end)

    it("returned parser has a parse method", function()
        local p = sql_parser.create_parser("SELECT * FROM t")
        assert.is_function(p.parse)
    end)

    it("parsing via create_parser returns same root rule_name as parse()", function()
        local src = "SELECT * FROM users"
        local ast1 = sql_parser.parse(src)
        local p    = sql_parser.create_parser(src)
        local ast2, err = p:parse()
        assert.is_nil(err)
        assert.are.equal(ast1.rule_name, ast2.rule_name)
    end)
end)

-- =========================================================================
-- Error handling
-- =========================================================================

describe("error handling", function()
    it("raises an error for completely invalid SQL", function()
        assert.has_error(function()
            sql_parser.parse("@@@ GARBAGE @@@")
        end)
    end)

    it("raises an error for incomplete SELECT (missing FROM)", function()
        assert.has_error(function()
            sql_parser.parse("SELECT name")
        end)
    end)

    it("raises an error for empty input", function()
        assert.has_error(function()
            sql_parser.parse("")
        end)
    end)
end)
